sql/index.go

```diff
@@ -39,16 +40,37 @@ type Index interface {
 	// IsGenerated returns whether this index was generated. Generated indexes
 	// are used for index access, but are not displayed (such as with SHOW INDEXES).
 	IsGenerated() bool
+	// ColumnExpressionTypes returns each expression and its associated Type. Each expression string should exactly
+	// match the string returned from Index.Expressions().
+	ColumnExpressionTypes(ctx *Context) []ColumnExpressionType
+}
+
+type LookupBuilderKey []interface{}
+
+// LookupPlaceholder is an integer corresponding to an index field in the
+// inner join table (left table / primary table)-provided LookupBuilderKey.
+type LookupPlaceholder int
+
+// ConditionalRange is a Range whose literal bound values are replaced
+// with LookupPlaceholder
+type ConditionalRange Range
+
+type LookupIndex interface {
 	// NewLookup returns a new IndexLookup for the ranges given. Ranges represent filters over columns. Each Range
 	// is ordered by the column expressions (as returned by Expressions) with the RangeColumnExpr representing the
 	// searchable area for each column expression. Each Range given will not overlap with any other ranges. Additionally,
 	// all ranges will have the same length, and may represent a partial index (matching a prefix rather than the entire
 	// index). If an integrator is unable to process the given ranges, then a nil may be returned. An error should be
 	// returned only in the event that an error occurred.
 	NewLookup(ctx *Context, ranges ...Range) (IndexLookup, error)
-	// ColumnExpressionTypes returns each expression and its associated Type. Each expression string should exactly
-	// match the string returned from Index.Expressions().
-	ColumnExpressionTypes(ctx *Context) []ColumnExpressionType
+	// WithConditionalRanges passes a set of range conditions for an IndexLookup
+	// whose literal values have been replaced with LookupPlaceholder.
+	WithConditionalRanges(ranges ...ConditionalRange) Index
+	// NewSecondaryLookup returns an IndexLookup constructed from a LookupBuilderKey
+	// from an inner join table and a previously cached conditional range.
+	//TODO: lookups do not currently support ranges, but if we do this may return
+	// an index with an invalid range lookup
+	NewSecondaryLookup(*Context, LookupBuilderKey) (IndexLookup, error)
```

Comment:
@andrew-wm-arthur 's feedback was maybe parametrized ranges should be cached on the GMS side, opaque to the integrator. The integrator should be given valid ranges and internally convert into an appropriate lookup. The tradeoff is shaving a few percent off the potential perf win.